Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm

```diff
@@ -6503,13 +6502,75 @@ - (void)_handleKeyUIEvent:(::UIEvent *)event
             _seenHardwareKeyDownInNonEditableElement = YES;
             [self reloadInputViews];
         }
-        [super _handleKeyUIEvent:event];
+        [super pressesBegan:presses withEvent:event];
         return;
     }
 
-    [super _handleKeyUIEvent:event];
+    [super pressesBegan:presses withEvent:event];
+}
+
+- (void)pressesEnded:(NSSet<UIPress *> *)presses withEvent:(UIPressesEvent *)event
+{
+    bool isHardwareKeyboardEvent = !!event._hidEvent;
+    // We only want to handle key event from the hardware keyboard when we are
+    // first responder and we are not interacting with editable content.
+    if ([self isFirstResponder] && isHardwareKeyboardEvent && (_inputPeripheral || !_page->editorState().isContentEditable)) {
+        if ([_inputPeripheral respondsToSelector:@selector(handleKeyEvent:)]) {
+            if ([_inputPeripheral handleKeyEvent:event])
+                return;
+        }
+        if (!_seenHardwareKeyDownInNonEditableElement) {
+            _seenHardwareKeyDownInNonEditableElement = YES;
+            [self reloadInputViews];
+        }
+        [super pressesEnded:presses withEvent:event];
+        return;
+    }
```

Comment:
It would be nice to share some of this logic with the other 3 `-presses*` methods as well — maybe with a helper method like `- (BOOL)_handlePressesEvent:(UIPressesEvent *)event;`.

Reply:
Yup, made a helper here too!

```
- (BOOL)_tryToHandlePressesEvent:(UIPressesEvent *)event;
```